rir/src/ir/Compiler.cpp

```diff
@@ -1073,6 +1073,87 @@ bool compileSpecialCall(Context& ctx, SEXP ast, SEXP fun, SEXP args_) {
     return false;
 }
 
+SEXP findRIRClosure(SEXP sym, SEXP rho) {
+    SEXP fun;
+    while (rho != R_EmptyEnv) {
+        fun = findVarInFrame3(rho, sym, TRUE);
+        if (fun != R_UnboundValue) {
+            if (TYPEOF(fun) == PROMSXP || TYPEOF(fun) == BUILTINSXP ||
+                    TYPEOF(fun) == SPECIALSXP || fun == R_MissingArg)
+                return nullptr;
+            if (isValidClosureSEXP(fun))
+                return fun;
+        }
+        rho = ENCLOS(rho);
+    }
+    return nullptr;
+}
+
+// try to look up a closure
+bool compileWithGuess(Context& ctx, SEXP ast, SEXP fun, SEXP args_) {
+    RList args(args_);
+    CodeStream& cs = ctx.cs();
+
+    SEXP cls = findRIRClosure(fun, R_GlobalEnv);  // or current env of the closure?
+    if (!cls)
+        return false;
+
+    RList formals(FORMALS(cls));
+    if (formals.length() != args.length() || args.length() == 0)
+        return false;
+
+    for (auto farg : formals)
+        if (farg == R_DotsSymbol)
+            return false;
+
+    for (auto a = args.begin(); a != args.end(); ++a) {
+        if (a.hasTag())
+            return false;
+        switch (TYPEOF(*a)) {
+        case NILSXP:
+        case LISTSXP:
+        case LGLSXP:
+        case INTSXP:
+        case REALSXP:
+        case STRSXP:
+        case CPLXSXP:
+        case RAWSXP:
+        case S4SXP:
+        case SPECIALSXP:
+        case BUILTINSXP:
+        case ENVSXP:
+        case CLOSXP:
+        case VECSXP:
+        case EXTPTRSXP:
+        case WEAKREFSXP:
+        case EXPRSXP:
+            break;
+        default:
+            return false;
+        }
+    }
+
+    // maybe use promise_ and compile everything either as push or
+    // as promise?
+
+    // do argument matching / shuffling?
+    // cannot reuse matchArgs from gnur (but could rewrite it)
+
+    cs << BC::guardName(fun, cls);
+
+    FunctionSignature* signature = new FunctionSignature();
+    signature->argsOnStack = true;
```

Comment:
Oh, that's why `argsOnStack` is public. But I'd probably add a new constructor for `FunctionSignature`. Unless we plan on mutating `argsOnStack` after construction?